Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseTeleOp.java

```diff
@@ -1,61 +1,206 @@
+        headingDegrees = currentAngle - startAngle;
+        negativeHeadingRadians = Math.toRadians(-headingDegrees);
+
+        if (gamepad1.left_trigger > 0.25) {
+            x = gamepad1.left_stick_x * 0.3;
+            y = -gamepad1.left_stick_y * 0.3;
+            t = gamepad1.right_stick_x * 0.2;
+        } else {
+            x = gamepad1.left_stick_x * 0.75;
+            y = -gamepad1.left_stick_y * 0.75;
+            t = gamepad1.right_stick_x * 0.5;
+        }
+
+        xRotatedVector = x * Math.cos(negativeHeadingRadians) - y * Math.sin(negativeHeadingRadians);
+        yRotatedVector = x * Math.sin(negativeHeadingRadians) + y * Math.cos(negativeHeadingRadians);
+
+        angleToCardinal = headingDegrees % 90;
+
+        if (angleToCardinal <= 10) {
+            t += angleToCardinal / 50.0;
+        } else if (angleToCardinal >= 80) {
+            t -= (90 - angleToCardinal) / 50.0;
+        }
+
+        // drive only left and right
+        if (Math.abs(Math.toDegrees(Math.atan(gamepad1.left_stick_y / gamepad1.left_stick_x))) <= 45) {
+            ratio = 1 / Math.max(Math.abs(xRotatedVector) + Math.abs(t), 1);
+
+            xPower = xRotatedVector * ratio;
+            tPower = t * ratio;
 
-        // case for driving the robot left and right
-        if (Math.abs(Math.toDegrees(Math.atan(gamepad1.left_stick_y / gamepad1.left_stick_x))) < Constants.DRIVE_DEADZONE_DEGREES) {
             driveWithIMU(xPower, 0.0, tPower);
 
-        // case for driving the robot forwards and backwards
-        } else if (Math.abs(Math.toDegrees(Math.atan(gamepad1.left_stick_y / gamepad1.left_stick_x))) > Constants.DRIVE_DEADZONE_DEGREES) {
+        // drive only forwards and backwards
+        } else if (Math.abs(Math.toDegrees(Math.atan(gamepad1.left_stick_y / gamepad1.left_stick_x))) >= 45) {
+            ratio = 1 / Math.max(Math.abs(yRotatedVector) + Math.abs(t), 1);
+
+            yPower = yRotatedVector * ratio;
+            tPower = t * ratio;
+
             driveWithIMU(0.0, yPower, tPower);
 
-        // case for if the deadzone limits are passed, the robot drives normally
+        // drive normally
         } else {
+            ratio = 1 / Math.max(Math.abs(xRotatedVector) + Math.abs(yRotatedVector) + Math.abs(t), 1);
+
+            xPower = xRotatedVector * ratio;
+            yPower = yRotatedVector * ratio;
+            tPower = t * ratio;
+
             driveWithIMU(xPower, yPower, tPower);
         }
     }
 
     public void driveGrabberWithController() {
-        if (gamepad2.a) {
-            driveGrabber(false);
-        } else if (gamepad2.x) {
-            driveGrabber(true);
+        if (gamepad2.x) {
+            driveGrabber(Constants.GRABBER_CLOSE_POSITION);
+        } else if (gamepad2.a) {
+            driveGrabber(Constants.GRABBER_OPEN_POSITION);
+        } else if (gamepad2.b) {
+            driveGrabber(Constants.GRABBER_INITIALIZE_POSITION);
         }
     }
 
     public void driveSlidesWithController() {
```

Comment:
This method needs more comments explaining what is going on.

Reply:
it just opens the grabber to a certain position, theoretically the constants variable names should be good enough?

Reply:
Theoretically, but I don't think it's very clear. Also you should have a javadoc comment for all public methods. I don't understand why you're comparing the two stacks against each other or the two junctions. Maybe your variable names aren't clear enough.

Reply:
The javadoc comments have been added to all the methods in baseTeleOp